### PR TITLE
Fix DEVSUP-749: Fix potential deadlock

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

3.8 backport: https://github.com/arangodb/arangodb/pull/14102
3.7 backport: https://github.com/arangodb/arangodb/pull/14101
3.6 backport: https://github.com/arangodb/arangodb/pull/14104

Fix issue DEVSUP-749: Fix potential deadlock when executing concurrent view/link DDL operations and index DDL operations on the same collection.

There is a potential issue with a recursive lock on indexes that does seem to have an effect when view/link DDL operations run concurrently to index DDL operations.

Steps to reproduce/test:
1) create collection with an arangosearch view, populate with 100k documents
2) start two shells that each add and then delete a link for the view, in an endless loop
3) start a shell that creates and drops a persistent index in an endless loop

For 1):
```
if (db._collection("testCollection") === null) {
  db._create("testCollection");
}
db._createView("testView", "arangosearch", {});

for (let i = 0; i < 100 * 1000; ++i) {
  db.testCollection.insert({ value1: i, value2: "test" + i });
}
```
For 2):
```
for (let i = 0; i < 10000000; ++i) {
  db._view("testView").properties({ links : { testCollection : null }}, true);
  db._view("testView").properties({ links : { testCollection : { includeAllFields: true}} }, true);
  print("ITERATION: ", i, db._view("testview"), db.testCollection.indexes());
}
```
For 3):
```
for (let i = 0; i < 10000000; ++i) {
  let idx = db.testCollection.ensureIndex({ type: "persistent", fields: ["value2"] });
  db.testCollection.dropIndex(idx); 
}
```

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: 3.7: https://github.com/arangodb/arangodb/pull/14101, 3.8: https://github.com/arangodb/arangodb/pull/14102, potentially 3.6

#### Related Information

- [x] GitHub issue / Jira ticket number: <https://arangodb.atlassian.net/browse/DEVSUP-749

### Testing & Verification

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*

### CODE REVIEW

**Comment by gnusi** [comment]
@jsteemann could we please add you test scenario into our js tests at least for `devel` and `3.8`?

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 1]
```diff
@@ -644,7 +660,7 @@ bool RocksDBCollection::dropIndex(IndexId iid) {
           LogicalDataSource::Serialization::PersistenceWithInProgress);
 
   // log this event in the WAL and in the collection meta-data
-  res = engine.writeCreateCollectionMarker(  // write marker
+  Result res = engine.writeCreateCollectionMarker(  // write marker
```
Comment (Dronplane) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:663]: It looks like here we will may get one dropped index  "resurrected" with the following sequence of events:
(2 threads are dropping unrelated indexes A and B)
1.  Both threads reaches line  598 (all lines are given in NEW version)
2. Thread 1 gets writelock and removes index A from _indexes. Thread 2 stays blocked
3.  Thread 1 reaches line 657 and gets slice without index A but with index B
4. Thread 1 is blocked and thread 2 starts.
5. Thread 2 removes index B from _indexes
6. Thread 2 also reaches line 657 (with old version Thread 2 would still wait on writelock until Thread 1 completes its write)
7. Thread 2 gets slice without index A and B and goes to line 663 and writes it
8. Thread 1 is unblocked and writes slice with index B  - index B is a zombie now

Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBCollection.cpp:663]: @Dronplane : true. I will try a different approach now.

**Comment by jsteemann** [comment]
@Dronplane : adjusted the fix. Can you check again? Thanks!

**Review by Dronplane — APPROVED** [review]
LGTM

File: lib/Basics/RecursiveLocker.h [hunk 2]
```diff
@@ -112,37 +123,42 @@ class RecursiveReadLocker {
...
+      // it is possible that we cannot acquire the lock, e.g.
+      // if we are a TRY locker
+      if (_locker.isLocked()) {
+        _owner.store(std::this_thread::get_id());
+        _update = owned;
+        _locked = true;
+      }
+    } else {
+      // we are the owning thread
+      _locked = true;
```
Comment (mpoeter) [inline lib/Basics/RecursiveLocker.h:160]:
```suggestion
      _owner.store(std::this_thread::get_id());
      _update = owned;
```

File: lib/Basics/RecursiveLocker.h [hunk 3]
```diff
@@ -36,70 +36,81 @@ namespace arangodb {
...
     }
```
Comment (mpoeter) [inline lib/Basics/RecursiveLocker.h:74]:
```suggestion
    }
    _locked = true;
```

File: lib/Basics/RecursiveLocker.h [hunk 4]
```diff
@@ -112,37 +123,42 @@ class RecursiveReadLocker {
...
```
Comment (mpoeter) [inline lib/Basics/RecursiveLocker.h:161]:
```suggestion
    }
    _locked = true;
```

**Review by mpoeter — CHANGES_REQUESTED** [review]

File: lib/Basics/RecursiveLocker.h [hunk 5]
```diff
@@ -62,44 +64,47 @@ class RecursiveMutexLocker {
       _owner.store(std::this_thread::get_id());
       _update = owned;
     }
+    _locked = true;
+    _locked = true;
```
Comment (mpoeter) [inline lib/Basics/RecursiveLocker.h:68]:
```suggestion
    _locked = true;
```

**Review by mpoeter — APPROVED** [review]

**Pull request merged**